File: spring-web-java/src/main/java/spring/web/java/application/service/OrderService.java
package spring.web.java.application.service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.ObjectUtils;

import lombok.RequiredArgsConstructor;
import spring.web.java.domain.model.entity.Item;
import spring.web.java.domain.model.entity.Member;
import spring.web.java.domain.model.entity.Order;
import spring.web.java.domain.model.entity.OrderItem;
import spring.web.java.domain.model.enums.OrderStatus;
import spring.web.java.domain.repository.ItemRepository;
import spring.web.java.domain.repository.MemberRepository;
import spring.web.java.domain.repository.OrderRepository;
import spring.web.java.presentation.dto.request.OrderItemSaveRequest;
import spring.web.java.presentation.dto.request.OrderSaveRequest;
import spring.web.java.presentation.dto.response.OrderResponse;
import spring.web.java.presentation.exception.EntityNotFoundException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class OrderService {

	private final MemberRepository memberRepository;
	private final ItemRepository itemRepository;
	private final OrderRepository orderRepository;

	@Transactional
	public OrderResponse saveOrder(OrderSaveRequest orderSaveRequest) {
		Long memberId = orderSaveRequest.memberId();
		Member member = memberRepository.findById(memberId)
			.orElseThrow(() -> new EntityNotFoundException(Member.class, memberId));

		List<OrderItem> orderItems = new ArrayList<>();
		if (!ObjectUtils.isEmpty(orderSaveRequest.orderItems())) {
			for (OrderItemSaveRequest requestOrderIterm : orderSaveRequest.orderItems()) {
				Long itemId = requestOrderIterm.itemId();
				Item item = itemRepository.findById(itemId)
					.orElseThrow(() -> new EntityNotFoundException(Item.class, itemId));

				orderItems.add(OrderItem.create(item, requestOrderIterm.count()));
			}
		}

		Order order = orderRepository.save(
			Order.create(member, orderItems)
		);

		return new OrderResponse(order);
	}

	public Page<OrderResponse> findOrders(
		Long memberId,
		OrderStatus orderStatus,
		Pageable pageable
	) {
		return orderRepository.findAll(pageable, memberId, orderStatus)
			.map(OrderResponse::new);
	}

	public OrderResponse findOrder(Long id) {
		Order order = orderRepository.findById(id)
			.orElseThrow(() -> new EntityNotFoundException(Order.class, id));

		return new OrderResponse(order);
	}

	@Transactional
	public OrderResponse cancelOrder(Long id) {
		Order order = orderRepository.findById(id)
			.orElseThrow(() -> new EntityNotFoundException(Order.class, id));

		order.cancel();

		return new OrderResponse(order);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/application/service/OrderService.kt
package spring.web.kotlin.application.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.model.entity.Member
import spring.web.kotlin.domain.model.entity.Order
import spring.web.kotlin.domain.model.entity.OrderItem
import spring.web.kotlin.domain.model.enums.OrderStatus
import spring.web.kotlin.domain.repository.ItemRepository
import spring.web.kotlin.domain.repository.MemberRepository
import spring.web.kotlin.domain.repository.OrderRepository
import spring.web.kotlin.presentation.dto.request.OrderSaveRequest
import spring.web.kotlin.presentation.dto.response.OrderResponse
import spring.web.kotlin.presentation.exception.EntityNotFoundException

@Service
@Transactional(readOnly = true)
class OrderService(
    private val memberRepository: MemberRepository,
    private val itemRepository: ItemRepository,
    private val orderRepository: OrderRepository
) {
    @Transactional
    fun saveOrder(orderSaveRequest: OrderSaveRequest): OrderResponse {
        val (memberId, city, street, zipcode, requestOrderItems) = orderSaveRequest

        val member = memberRepository.findByIdOrNull(memberId)
            ?: throw EntityNotFoundException(Member::class.java, memberId)

        val orderItems = requestOrderItems.map { orderItem ->
            val item = itemRepository.findByIdOrNull(orderItem.itemId)
                ?: throw EntityNotFoundException(OrderItem::class.java, orderItem.itemId)

            OrderItem.create(item, orderItem.count)
        }

        val order = orderRepository.save(
            Order.create(
                member = member,
                orderItems = orderItems,
            )
        )

        return OrderResponse(order)
    }

    fun findOrders(
        memberId: Long?,
        orderStatus: OrderStatus?,
        pageable: Pageable
    ): Page<OrderResponse> {
        return orderRepository.findAll(pageable, memberId, orderStatus)
            .map(::OrderResponse)
    }

    fun findOrder(id: Long): OrderResponse {
        val order = orderRepository.findByIdOrNull(id)
            ?: throw EntityNotFoundException(Order::class.java, id)

        return OrderResponse(order)
    }

    @Transactional
    fun cancelOrder(id: Long): OrderResponse {
        val order = orderRepository.findByIdOrNull(id)
            ?: throw EntityNotFoundException(Order::class.java, id)

        order.cancel()

        return OrderResponse(order)
    }
}